man/ibroute.8
.TH IBROUTE 8 "July 25, 2006" "OpenIB" "OpenIB Diagnostics"

.SH NAME
ibroute \- query InfiniBand switch forwarding tables

.SH SYNOPSIS
.B ibroute
[\-d(ebug)] [-a(ll)] [-n(o_dests)] [-v(erbose)] [\-D(irect)] [\-G(uid)]
[-M(ulticast)] [-s smlid] [\-C ca_name] [\-P ca_port]
[\-t(imeout) timeout_ms] [\-V(ersion)] [\-h(elp)]
[<dest dr_path|lid|guid> [<startlid> [<endlid>]]]

.SH DESCRIPTION
.PP
ibroute uses SMPs to display the forwarding tables (unicast
(LinearForwardingTable or LFT) or multicast (MulticastForwardingTable or MFT))
for the specified switch LID and the optional lid (mlid) range.
The default range is all valid entries in the range 1...FDBTop.

.SH OPTIONS

.PP
.TP
\fB\-a\fR, \fB\-\-all\fR
show all lids in range, even invalid entries
.TP
\fB\-n\fR, \fB\-\-no_dests\fR
do not try to resolve destinations
.TP
\fB\-M\fR, \fB\-\-Multicast\fR
show multicast forwarding tables
In this case, the range parameters are specifying the mlid range.

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-d      raise the IB debugging level.
        May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
        May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Addressing flags
.PP
\-D      use directed path address arguments. The path
        is a comma separated list of out ports.
        Examples:
        "0"             # self port
        "0,1,2,1,4"     # out via port 1, then 2, ...
.PP
\-G      use GUID address argument. In most cases, it is the Port GUID.
        Example:
        "0x08f1040023"
.PP
\-s <smlid>      use 'smlid' as the target lid for SM/SA queries.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.

Multiple CA/Multiple Port Support

When no IB device or port is specified, the port to use is selected
by the following criteria:
.PP
1. the first port that is ACTIVE.
.PP
2. if not found, the first port that is UP (physical link up).

If a port and/or CA name is specified, the user request is
attempted to be fulfilled, and will fail if it is not possible.

.SH EXAMPLES

.PP
Unicast examples
.PP
ibroute 4       # dump all lids with valid out ports of switch with lid 4
.PP
ibroute -a 4    # same, but dump all lids, even with invalid out ports
.PP
ibroute -n 4    # simple dump format - no destination resolution
.PP
ibroute 4 10    # dump lids starting from 10 (up to FDBTop)
.PP
ibroute 4 0x10 0x20     # dump lid range
.PP
ibroute -G 0x08f1040023 # resolve switch by GUID
.PP
ibroute -D 0,1  # resolve switch by direct path

.PP
Multicast examples
.PP
ibroute -M 4    # dump all non empty mlids of switch with lid 4
.PP
ibroute -M 4 0xc010 0xc020      # same, but with range
.PP
ibroute -M -n 4 # simple dump format

.SH SEE ALSO
.BR ibtracert (8)

.SH AUTHOR
.TP
Hal Rosenstock
.RI < halr@voltaire.com >


ChangeLog
commit 596aa88065f89906884a821a1bb4eb631abd6af0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:57:58 2009 -0500

    Spin release version 1.2.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0ec532c2b6b92be6adf58849c9e3d215a50d306d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:52:51 2009 -0500

    Support new device ids: 0x35, 0x36.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f685c8fe7e77e64614d825e563dd9f02a0b1ae16
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:46 2008 -0500

    Spin release 1.2.2
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ebd19d077d331270805aefa9ca00d3661561ba87
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:27 2008 -0500

    Relax the required firmware level.
    
    Turns out that libcxgb3 can really run on either 6.0 or 5.0
    firmware.  So I'm relaxing the requirement...
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6f7485feb244d8571fcab2292ef92c97bea48df0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue May 13 09:09:35 2008 -0500

    Spin release 1.2.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 309916b23ba384a0bd297e7e51b6b89bff357a78
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:45:01 2008 -0500

    Use local stag == 1 to silently drop the p2p read response.
    
    The current code fails if there are read completions -and-
    the p2p read response in the cq when its flushed.
    
    This fix uses the fact that the RTR read is to target stag 1,
    which is not a valid stag.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 67d519c4a9e28e54a9e436404156c9ae7eb60314
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:25:41 2008 -0500

    Handle wrapping when flushing the software sq.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a3dcea751a94ad5e0a9cad6c55a839a9848f327c
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Spin 1.2.0 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit cc2ef741842883233a77e18d65a828c711e581a3
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit 92322b93f48fa98b4fadb736ae49dafda237272a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Apr 24 14:23:50 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit a8fb6221af4d914f39c1e71f9331f6d428d698c2
Author: Steve Wise (Open Grid Computing) <swise@hosting.openfabrics.org>
Date:   Tue Apr 1 15:30:11 2008 -0700

    Spin 1.1.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit db41ebc10f4cc4c2fe7826589636ab7d4b804218
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 26 08:27:12 2008 -0600

    Provide an empty stub for iwch_query_qp.
    
    This prevents crashes when calling ibv_query_qp due to an expected
    library provided function.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ec4f7f1c6b8beceb9001519bc4e2ee89fbb2e6d0
Author: Dotan Barak <dotanb@dev.mellanox.co.il>
Date:   Tue Feb 26 08:23:36 2008 -0600

    Remove duplicated line.
    
    Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8218482d03898f859e1fa272649fafc64f91f05f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Feb 12 15:16:03 2008 -0600

    Spin release 1.1.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 1bc6930d8884646ef73db71a95cc32c96c20ba32
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 12 14:46:10 2008 -0600

    libcxgb3: zeroing of wc_flags
    
    The wc_flags field in struct ibv_wc is left uninitialized in
    iwch_poll_cq_one.  User space applications may check this field and
    deterministically perform actions based on the garbage in the field.
    Zeroing this out will prevent this unintended behavior.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 946e1572d38d786038d3bca5a2ec184dcf7636c4
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jan 25 16:47:01 2008 -0600

    Spin release 1.1.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 94fb771a44212a679d7f83ee69df24e841e0ce08
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jan 25 13:04:31 2008 -0600

    Fix path for obtaining the firmware version.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 4bbb15965b53cad6273094e95b06895010ca025c
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Wed Jan 23 09:40:28 2008 -0600

    Non-supported functions should return NULL when returning pointers.
    
    Some/Most user space programs will not check for a (void *) to -ENOSYS,
    which can look like a real address until referenced.
    
    This patch converts the uses of (void *) -ENOSYS to NULL.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 10893a7a45e6913483d90023bb90bf9cb2420384
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jan 3 17:19:27 2008 -0600

    Spin release 1.1.2.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a35c74c32d038bd7508ad3cb6c5291c1865d3ba4
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jan 3 17:16:27 2008 -0600

    Fix compiler warnings on 32b.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ef3b9d4f99eb6e18c29b70599edf99d0715f03fb

src/ibroute.c
/*
 * Copyright (c) 2004-2008 Voltaire Inc.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <inttypes.h>
#include <getopt.h>
#include <netinet/in.h>

#include <infiniband/umad.h>
#include <infiniband/mad.h>
#include <complib/cl_nodenamemap.h>

#include "ibdiag_common.h"

struct ibmad_port *srcport;

static int brief, dump_all, multicast;

/*******************************************/

char *
check_switch(ib_portid_t *portid, unsigned int *nports, uint64_t *guid,
	     uint8_t *sw, char *nd)
{
	uint8_t ni[IB_SMP_DATA_SIZE] = {0};
	int type;

	DEBUG("checking node type");
	if (!smp_query_via(ni, portid, IB_ATTR_NODE_INFO, 0, 0, srcport)) {
		xdump(stderr, "nodeinfo\n", ni, sizeof ni);
		return "node info failed: valid addr?";
	}

	if (!smp_query_via(nd, portid, IB_ATTR_NODE_DESC, 0, 0, srcport))
		return "node desc failed";

	mad_decode_field(ni, IB_NODE_TYPE_F, &type);
	if (type != IB_NODE_SWITCH)
		return "not a switch";

	DEBUG("Gathering information about switch");
	mad_decode_field(ni, IB_NODE_NPORTS_F, nports);
	mad_decode_field(ni, IB_NODE_GUID_F, guid);

	if (!smp_query_via(sw, portid, IB_ATTR_SWITCH_INFO, 0, 0, srcport))
		return "switch info failed: is a switch node?";

	return 0;
}

#define IB_MLIDS_IN_BLOCK	(IB_SMP_DATA_SIZE/2)

int
dump_mlid(char *str, int strlen, unsigned mlid, unsigned nports,
	  uint16_t mft[16][IB_MLIDS_IN_BLOCK])
{
	uint16_t mask;
	unsigned i, chunk, bit, nonzero = 0;

	if (brief) {
		int n = 0;
		unsigned chunks = ALIGN(nports + 1, 16) / 16;
		for (i = 0; i < chunks; i++) {
			mask = ntohs(mft[i][mlid%IB_MLIDS_IN_BLOCK]);
			if (mask)
				nonzero++;
			n += snprintf(str + n, strlen - n, "%04hx", mask);
			if (n >= strlen) {
				n = strlen;
				break;
			}
		}
		if (!nonzero && !dump_all) {
			str[0] = 0;
			return 0;
		}
		return n;
	}
	for (i = 0; i <= nports; i++) {
		chunk = i / 16;
		bit = i % 16;

		mask = ntohs(mft[chunk][mlid%IB_MLIDS_IN_BLOCK]);
		if (mask)
			nonzero++;
		str[i*2] = (mask & (1 << bit)) ? 'x' : ' ';
		str[i*2+1] = ' ';
	}
	if (!nonzero && !dump_all) {
		str[0] = 0;
		return 0;
	}
	str[i*2] = 0;
	return i * 2;
}

uint16_t mft[16][IB_MLIDS_IN_BLOCK];

char *
dump_multicast_tables(ib_portid_t *portid, unsigned startlid, unsigned endlid)
{
	char nd[IB_SMP_DATA_SIZE] = {0};
	uint8_t sw[IB_SMP_DATA_SIZE] = {0};
	char str[512];
	char *s;
	uint64_t nodeguid;
	uint32_t mod;
	unsigned block, i, j, e, nports, cap, chunks, startblock, lastblock;
	int n = 0;

	if ((s = check_switch(portid, &nports, &nodeguid, sw, nd)))
		return s;

	mad_decode_field(sw, IB_SW_MCAST_FDB_CAP_F, &cap);

	if (!endlid || endlid > IB_MIN_MCAST_LID + cap - 1)
		endlid = IB_MIN_MCAST_LID + cap - 1;

	if (!startlid)
		startlid = IB_MIN_MCAST_LID;

	if (startlid < IB_MIN_MCAST_LID) {
		IBWARN("illegal start mlid %x, set to %x", startlid, IB_MIN_MCAST_LID);
		startlid = IB_MIN_MCAST_LID;
	}

	if (endlid > IB_MAX_MCAST_LID) {
		IBWARN("illegal end mlid %x, truncate to %x", endlid, IB_MAX_MCAST_LID);
		endlid = IB_MAX_MCAST_LID;
	}

	printf("Multicast mlids [0x%x-0x%x] of switch %s guid 0x%016" PRIx64 " (%s):\n",
		startlid, endlid, portid2str(portid), nodeguid, clean_nodedesc(nd));

	if (brief)
		printf(" MLid       Port Mask\n");
	else {
		if (nports > 9) {
			for (i = 0, s = str; i <= nports; i++) {
				*s++ = (i%10) ? ' ' : '0' + i/10;
				*s++ = ' ';
			}
			*s = 0;
			printf("            %s\n", str);
		}
		for (i = 0, s = str; i <= nports; i++)
			s += sprintf(s, "%d ", i%10);
		printf("     Ports: %s\n", str);
		printf(" MLid\n");
	}
	if (ibverbose)
		printf("Switch multicast mlid capability is %d\n", cap);

	chunks = ALIGN(nports + 1, 16) / 16;

	startblock = startlid / IB_MLIDS_IN_BLOCK;
	lastblock = endlid / IB_MLIDS_IN_BLOCK;
	for (block = startblock; block <= lastblock; block++) {
		for (j = 0; j < chunks; j++) {
			mod = (block - IB_MIN_MCAST_LID/IB_MLIDS_IN_BLOCK) | (j << 28);

			DEBUG("reading block %x chunk %d mod %x", block, j, mod);
			if (!smp_query_via(mft + j, portid,
					IB_ATTR_MULTICASTFORWTBL, mod, 0, srcport))
				return "multicast forwarding table get failed";
		}

		i = block * IB_MLIDS_IN_BLOCK;
		e = i + IB_MLIDS_IN_BLOCK;
		if (i < startlid)
			i = startlid;
		if (e > endlid + 1)
			e = endlid + 1;

		for (; i < e; i++) {
			if (dump_mlid(str, sizeof str, i, nports, mft) == 0)
				continue;
			printf("0x%04x      %s\n", i, str);
			n++;
		}
	}

	printf("%d %smlids dumped \n", n, dump_all ? "" : "valid ");
	return 0;
}

int
dump_lid(char *str, int strlen, int lid, int valid)
{
	char nd[IB_SMP_DATA_SIZE] = {0};
	uint8_t ni[IB_SMP_DATA_SIZE] = {0};
	uint8_t pi[IB_SMP_DATA_SIZE] = {0};
	ib_portid_t lidport = {0};
	static int last_port_lid, base_port_lid;
	char ntype[50], sguid[30], desc[64];
	static uint64_t portguid;
	int baselid, lmc, type;

	if (brief) {
		str[0] = 0;
		return 0;
	}

	if (lid <= last_port_lid) {
		if (!valid)
			return snprintf(str, strlen, ": (path #%d - illegal port)",
					lid - base_port_lid);
		else if (!portguid)
			return snprintf(str, strlen,
					": (path #%d out of %d)",
					lid - base_port_lid + 1,
					last_port_lid - base_port_lid + 1);
		else {
			return snprintf(str, strlen,
					": (path #%d out of %d: portguid %s)",
					lid - base_port_lid + 1,
					last_port_lid - base_port_lid + 1,
					mad_dump_val(IB_NODE_PORT_GUID_F, sguid, sizeof sguid, &portguid));
		}
	}

	if (!valid)
		return snprintf(str, strlen, ": (illegal port)");

	portguid = 0;
	lidport.lid = lid;

	if (!smp_query_via(nd, &lidport, IB_ATTR_NODE_DESC, 0, 100, srcport) ||
	    !smp_query_via(pi, &lidport, IB_ATTR_PORT_INFO, 0, 100, srcport) ||
	    !smp_query_via(ni, &lidport, IB_ATTR_NODE_INFO, 0, 100, srcport))
		return snprintf(str, strlen, ": (unknown node and type)");

	mad_decode_field(ni, IB_NODE_PORT_GUID_F, &portguid);
	mad_decode_field(ni, IB_NODE_TYPE_F, &type);

	mad_decode_field(pi, IB_PORT_LID_F, &baselid);
	mad_decode_field(pi, IB_PORT_LMC_F, &lmc);

	if (lmc > 0) {
		base_port_lid = baselid;
		last_port_lid = baselid + (1 << lmc) - 1;
	}

	return snprintf(str, strlen, ": (%s portguid %s: %s)",
		mad_dump_val(IB_NODE_TYPE_F, ntype, sizeof ntype, &type),
		mad_dump_val(IB_NODE_PORT_GUID_F, sguid, sizeof sguid, &portguid),
		mad_dump_val(IB_NODE_DESC_F, desc, sizeof desc, clean_nodedesc(nd)));
}

char *
dump_unicast_tables(ib_portid_t *portid, int startlid, int endlid)
{
	char lft[IB_SMP_DATA_SIZE];
	char nd[IB_SMP_DATA_SIZE];
	uint8_t sw[IB_SMP_DATA_SIZE];
	char str[200], *s;
	uint64_t nodeguid;
	int block, i, e, top;
	unsigned nports;
	int n = 0, startblock, endblock;

	if ((s = check_switch(portid, &nports, &nodeguid, sw, nd)))
		return s;

	mad_decode_field(sw, IB_SW_LINEAR_FDB_TOP_F, &top);

	if (!endlid || endlid > top)
		endlid = top;

	if (endlid > IB_MAX_UCAST_LID) {
		IBWARN("illegal lft top %d, truncate to %d", endlid, IB_MAX_UCAST_LID);
		endlid = IB_MAX_UCAST_LID;
	}

	printf("Unicast lids [0x%x-0x%x] of switch %s guid 0x%016" PRIx64 " (%s):\n",
		startlid, endlid, portid2str(portid), nodeguid, clean_nodedesc(nd));

	DEBUG("Switch top is 0x%x\n", top);

	printf("  Lid  Out   Destination\n");
	printf("       Port     Info \n");
	startblock = startlid / IB_SMP_DATA_SIZE;
	endblock = ALIGN(endlid, IB_SMP_DATA_SIZE) / IB_SMP_DATA_SIZE;
	for (block = startblock; block <= endblock; block++) {
		DEBUG("reading block %d", block);
		if (!smp_query_via(lft, portid, IB_ATTR_LINEARFORWTBL, block,
				0, srcport))
			return "linear forwarding table get failed";
		i = block * IB_SMP_DATA_SIZE;
		e = i + IB_SMP_DATA_SIZE;
		if (i < startlid)
			i = startlid;
		if (e > endlid + 1)
			e = endlid + 1;

		for (;i < e; i++) {
			unsigned outport = lft[i % IB_SMP_DATA_SIZE];
			unsigned valid = (outport <= nports);

			if (!valid && !dump_all)
				continue;
			dump_lid(str, sizeof str, i, valid);
			printf("0x%04x %03u %s\n", i, outport & 0xff, str);
			n++;
		}
	}

	printf("%d %slids dumped \n", n, dump_all ? "" : "valid ");
	return 0;
}

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 'a':
		dump_all++;
		break;
	case 'M':
		multicast++;
		break;
	case 'n':
		brief++;
		break;
	default:
		return -1;
	}
	return 0;
}

int main(int argc, char **argv)
{
	int mgmt_classes[3] = {IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS};
	ib_portid_t portid = {0};
	unsigned startlid = 0, endlid = 0;
	char *err;

	const struct ibdiag_opt opts[] = {
		{ "all", 'a', 0, NULL, "show all lids, even invalid entries" },
		{ "no_dests", 'n', 0, NULL, "do not try to resolve destinations" },
		{ "Multicast", 'M', 0, NULL, "show multicast forwarding tables" },
		{ 0 }
	};
	char usage_args[] = "[<dest dr_path|lid|guid> [<startlid> [<endlid>]]]";
	const char *usage_examples[] = {
		" -- Unicast examples:",
		"4\t# dump all lids with valid out ports of switch with lid 4",
		"-a 4\t# same, but dump all lids, even with invalid out ports",
		"-n 4\t# simple dump format - no destination resolving",
		"4 10\t# dump lids starting from 10",
		"4 0x10 0x20\t# dump lid range",
		"-G 0x08f1040023\t# resolve switch by GUID",
		"-D 0,1\t# resolve switch by direct path",
		" -- Multicast examples:",
		"-M 4\t# dump all non empty mlids of switch with lid 4",
		"-M 4 0xc010 0xc020\t# same, but with range",
		"-M -n 4\t# simple dump format",
		NULL,
	};

	ibdiag_process_opts(argc, argv, NULL, NULL, opts, process_opt,
			    usage_args, usage_examples);

	argc -= optind;
	argv += optind;

	if (!argc)
		ibdiag_show_usage();

	if (argc > 1)
		startlid = strtoul(argv[1], 0, 0);
	if (argc > 2)
		endlid = strtoul(argv[2], 0, 0);

	srcport = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 3);
	if (!srcport)
		IBERROR("Failed to open '%s' port '%d'", ibd_ca, ibd_ca_port);

	if (!argc) {
		if (ib_resolve_self_via(&portid, 0, 0, srcport) < 0)
			IBERROR("can't resolve self addr");
	} else if (ib_resolve_portid_str_via(&portid, argv[0], ibd_dest_type,
			ibd_sm_id, srcport) < 0)
		IBERROR("can't resolve destination port %s", argv[1]);

	if (multicast)
		err = dump_multicast_tables(&portid, startlid, endlid);
	else
		err = dump_unicast_tables(&portid, startlid, endlid);

	if (err)
		IBERROR("dump tables: %s", err);

	mad_rpc_close_port(srcport);
	exit(0);
}
